### ISSUE TITLE
Completion source for vimscript

### ISSUE THREAD

**Issue by lervag** [roxma/nvim-completion-manager#70]
I've just recently started to look into NCM, and it seems quite good. One thing I miss from neocomplete which I used earlier is a good source for vimscript. Would it be possible to implement something similar through the use of [neco-vim](https://github.com/Shougo/neco-vim)?

**Comment by roxma** [comment]
>  Would it be possible to implement something similar through the use of neco-vim?

I is definitely possible.

There's a minimal example on how to implement a source with vimscript, `:help NCM-source-examples`. 

Use vim's help system for more information on NCM's api, for example, `:help cm#complete()`

Feel free to ask questions if the documentation is confusing.

**Comment by lervag** [comment]
Thanks. I'm trying to implement this now. I've first added this to my own `ftplugin/vim.vim` file:

```vim
augroup cm_complete_vim
  autocmd!
  autocmd User CmSetup call cm#register_source({
        \ 'name' : 'vim',
        \ 'priority': 9,
        \ 'scoping': 1,
        \ 'scopes': ['vim'],
        \ 'cm_refresh': 'personal#vim#cm_refresh',
        \ 'cm_refresh_patterns': ['fo$'],
        \ })
augroup END
```

The function `personal#vim#cm_refresh(opt, ctx)` is defined as follows (obv. in my own `autoload/personal/vim.vim` file):

```vim
function! personal#vim#cm_refresh(opt, ctx)
  let l:matches = ['foobar', 'foobaz', 'barbaz']
  call cm#complete(a:opt['name'], a:ctx, a:ctx['startcol'], l:matches)
endfunction
```

I would expect this to complete `foobar` and `foobaz` similar to the minimal source example. However, this does not work. Perhaps I'm missing something crucial here?

Also, I'm wondering, how would I debug this? Do you have any functions that shows which sources are available/loaded? I couldn't find anything like that in the docs, only `cm#get_context` (which might be more relevant later when I work on the actual completion through the `neco-vim` plugin).

**Comment by lervag** [comment]
Ok, if I call the `register_source` function without the autocmd, then it works. So, it seems the CmSetup fires before the FileType autocmd, or something like that. Do you have any recommendations wrt best practices here?

I also discovered the `g:_cm_sources` variable, so that provided the other part.

With that, I'm ready to try to couple this with neco-vim.

If I understand correctly, you prefer that sources are provided as external plugins. Thus, I should either make this as a PR to neco-vim, or provide a new plugin with an NCM source on top of neco-vim. Any comments on this?

**Comment by lervag** [comment]
I'm getting closer now, but it's a bit tricky to understand how things work. I'm using this:

```vim
call cm#register_source({
      \ 'name': 'vim',
      \ 'priority': 9,
      \ 'scoping': 1,
      \ 'scopes': ['vim'],
      \ 'cm_refresh': 'personal#vim#cm_refresh',
      \ 'cm_refresh_patterns': ['\w\+\.$'],
      \ })
```

Together with this:

```vim
function! personal#vim#cm_refresh(opt, ctx)
  let l:pos = necovim#get_complete_position(a:ctx.typed)
  let l:complete_str = strpart(a:ctx.typed, l:pos)
  let l:cnd = necovim#gather_candidates(a:ctx.typed, l:complete_str)

  call cm#complete(a:opt['name'], a:ctx, a:ctx['startcol'], l:cnd)
endfunction
```

You probably don't know the necovim api, but perhaps you understand more still. In particular, I am not sure what to pass to the `necovim#gather_candidates`. It asks for arguments `input` and `complete_str`. This version above seems to work for some things, such as autoload functions and intrinsic functions, but not for others such as options, dictionaries, etc.

Perhaps @Shougo would be able to help as well? I think to make this work properly, I would need to be able to translate from NCMs context to the context that is used by neocomplete/deoplete.

**Comment by Shougo** [comment]
> Perhaps @Shougo would be able to help as well? I think to make this work properly, I would need to be able to translate from NCMs context to the context that is used by neocomplete/deoplete.

Hm.  If you send PR, I will include it.

**Comment by roxma** [comment]
@lervag 

> So, it seems the CmSetup fires before the FileType autocmd, or something like that. Do you have any recommendations wrt best practices here?

Currently `CmSetup` is triggered on `BufWinEnter` only once. It should be registered at vim's startup(in `.vimrc` file, or `plugin/` directory). I have a plan to improve this behavior. But it is low priority.

If you put `register_source` call into filetype, there's no need to use `CmSetup`.

> If I understand correctly, you prefer that sources are provided as external plugins. Thus, I should either make this as a PR to neco-vim, or provide a new plugin with an NCM source on top of neco-vim. Any comments on this?

I do prefer sources provided as external plugins. I may possibly  include a source if it's not language specific or for a popular language that I'm able to mantain, and it should be easy to setup without dependencies.

I'll add it to the sources list either way.

> This version above seems to work for some things, such as autoload functions and intrinsic functions, but not for others such as options, dictionaries, etc.

I just tried deoplete version of necovim, there was no completion for vim options as well. There's completoin from current buffer if you type the option string twice.

**Comment by lervag** [comment]
Ok, thanks. I've opened a PR with `necovim` now.

**Comment by lervag** [comment]
FYI: `neco-vim` now has support for nvim-completion-manager for completing vimscript.
